### PR TITLE
Layout handling changes for the Railo admin

### PR DESCRIPTION
**Pull request by paulklinkenberg**
Hi,
I was about to start working on the upload functionality for Extensions, when I once again noticed the inline images and inline styles/scripts in the admin, and the .cfm extensions for javascript files.
Since the .cfm extensions will always reload on every request, and the inline images make the page unnecessarily heavy, I made some changes.

File getfile.cfm first does a security check for valid file path, and then either returns a 304 Not Modified based on ETag, or the file (img, css, js) contents plus an ETag header.
If getfile.cfm is run from /railo-context/getfile.cfm, it will also append 2 caching headers, which should cache the contents for 1 full day. This way, while developing locally, the files will renew, but when released, the files will be renewed only once a day.

Then, I added the following directories, and moved inline styles and javascripts to the respective directories:
/railo-admin/img
/railo-admin/js
/railo-admin/css

Kind regards,

Paul Klinkenberg

### CODE REVIEW

**Comment by micstriit** [comment]
In older versions the images was in external files and still is for older ie (there is a switch), but with extension .cfm. Problem with none cfm files is that they cannot be part of the ra, file and they are not reconized by the server. Because the Railo-context mapping is only know by the cfmlservlet, as long you not install the Railo default servlet. What never is by default.
In my test the internal way was the best solution. 

/micha

Vo mim iPhone gschickt

Am 31.03.2012 um 11:50 schrieb Paul Klinkenbergreply@reply.github.com:

**Comment by paulklinkenberg** [comment]
Hi Micha,

So, do you mean that a file "something.gif" cannot be inside a .ra file?
If it can't, then that's the end of the story.
But otherwise, it will work. I have changed the code so it uses 

So, all calls for the css/js/img files will still be calls to a .cfm file.
That will work I suspect?

Kind regards,

Paul Klinkenberg

Op 31 mrt 2012, om 13:10 heeft Michael Offner het volgende geschreven:

In older versions the images was in external files and still is for older ie (there is a switch), but with extension .cfm. Problem with none cfm files is that they cannot be part of the ra, file and they are not reconized by the server. Because the Railo-context mapping is only know by the cfmlservlet, as long you not install the Railo default servlet. What never is by default.
In my test the internal way was the best solution. 

/micha

Vo mim iPhone gschickt

Am 31.03.2012 um 11:50 schrieb Paul Klinkenbergreply@reply.github.com:

> Hi,
> I was about to start working on the upload functionality for Extensions, when I once again noticed the inline images and inline styles/scripts in the admin, and the .cfm extensions for javascript files.
> Since the .cfm extensions will always reload on every request, and the inline images make the page unnecessarily heavy, I made some changes.
> 
> File getfile.cfm first does a security check for valid file path, and then either returns a 304 Not Modified based on ETag, or the file (img, css, js) contents plus an ETag header.
> If getfile.cfm is run from /railo-context/getfile.cfm, it will also append 2 caching headers, which should cache the contents for 1 full day. This way, while developing locally, the files will renew, but when released, the files will be renewed only once a day.
> 
> Then, I added the following directories, and moved inline styles and javascripts to the respective directories:
> /railo-admin/img
> /railo-admin/js
> /railo-admin/css
> 
> Kind regards,
> 
> Paul Klinkenberg
> 
> You can merge this Pull Request by running:
> 
> git pull https://github.com/paulklinkenberg/railo develop-admin-changes
> 
> Or you can view, comment on it, or merge it online at:
> 
> https://github.com/getrailo/railo/pull/35
> 
> -- Commit Summary --
> - Merge remote branch 'Railo/develop' into develop
> - Removed unused old css file from the admin
> - Added getfile.cfm in the admin, which replaces inline images and inline style/javascript.
> 
> -- File Changes --
> 
> M railo-cfml/railo-admin/admin/admin_layout.cfm (374)
> D railo-cfml/railo-admin/admin/resources/admin.css.cfm (74)
> M railo-cfml/railo-admin/admin/web.cfm (140)
> A railo-cfml/railo-admin/css/style.css (90)
> A railo-cfml/railo-admin/getfile.cfm (56)
> A railo-cfml/railo-admin/img/arrow-active.gif (0)
> A railo-cfml/railo-admin/img/arrow-down.gif (0)
> A railo-cfml/railo-admin/img/arrow-left.gif (0)
> A railo-cfml/railo-admin/img/arrow-right.gif (0)
> A railo-cfml/railo-admin/img/arrow-up.gif (0)
> A railo-cfml/railo-admin/img/arrow.gif (0)
> A railo-cfml/railo-admin/img/back-left.png (0)
> A railo-cfml/railo-admin/img/bg-left.png (0)
> A railo-cfml/railo-admin/img/box-bg.png (0)
> A railo-cfml/railo-admin/img/box-del.png (0)
> A railo-cfml/railo-admin/img/box-left.png (0)
> A railo-cfml/railo-admin/img/box-right.png (0)
> A railo-cfml/railo-admin/img/debug_execute.gif (0)
> A railo-cfml/railo-admin/img/debug_minus.gif (0)
> A railo-cfml/railo-admin/img/debug_plus.gif (0)
> A railo-cfml/railo-admin/img/edit.png (0)
> A railo-cfml/railo-admin/img/input-button.png (0)
> A railo-cfml/railo-admin/img/input-shadow-error.png (0)
> A railo-cfml/railo-admin/img/input-shadow.png (0)
> A railo-cfml/railo-admin/img/left-tab-server.png (0)
> A railo-cfml/railo-admin/img/left-tab-web.png (0)
> A railo-cfml/railo-admin/img/left.png (0)
> A railo-cfml/railo-admin/img/maxl.png (0)
> A railo-cfml/railo-admin/img/maxr.png (0)
> A railo-cfml/railo-admin/img/minl.png (0)
> A railo-cfml/railo-admin/img/minus.png (0)
> A railo-cfml/railo-admin/img/plus.png (0)
> A railo-cfml/railo-admin/img/right-tab-server.png (0)
> A railo-cfml/railo-admin/img/right-tab-web.png (0)
> A railo-cfml/railo-admin/img/server-back.png (0)
> A railo-cfml/railo-admin/img/server-bgcolor.gif (0)
> A railo-cfml/railo-admin/img/server-minus.gif (0)
> A railo-cfml/railo-admin/img/server-railo-small.png (0)
> A railo-cfml/railo-admin/img/server-railo.png (0)
> A railo-cfml/railo-admin/img/shadow-left.gif (0)
> A railo-cfml/railo-admin/img/shadow-right.gif (0)
> A railo-cfml/railo-admin/img/tp.gif (0)
> A railo-cfml/railo-admin/img/web-back.png (0)
> A railo-cfml/railo-admin/img/web-bgcolor.gif (0)
> A railo-cfml/railo-admin/img/web-minus.gif (0)
> A railo-cfml/railo-admin/img/web-railo-small.png (0)
> A railo-cfml/railo-admin/img/web-railo.png (0)
> A railo-cfml/railo-admin/js/admin.js (103)
> A railo-cfml/railo-admin/js/jquery-1.7.2.min.js (4)
> A railo-cfml/railo-admin/js/jquery.blockUI.js (499)
> 
> -- Patch Links --
> 
> https://github.com/getrailo/railo/pull/35.patch
> https://github.com/getrailo/railo/pull/35.diff
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/getrailo/railo/pull/35

---

**Comment by micstriit** [comment]
Yes, a ra file can only contain cfm files. The old version works the same way as your solution but the gifs/pngs are stored in cfm files and can be loaded as include. All this exists and can be enabled by removing a single if statement. But in my tests this was slower and produced a lot of requests. From my point of view we should do a single image that contains all images as collage. Then you define a range in this big image to get a certain image. I have begin with that but run into problems, I have send a mail to the team list and still waiting for feedback ;-)

/Micha

Von meinem iPad gesendet

Am 31.03.2012 um 16:28 schrieb Paul Klinkenbergreply@reply.github.com:

**Comment by paulklinkenberg** [comment]
Hi Micha,

I think/hope you understood my solution wrong.
When I look at the WEB-INF directory of any web context, I see for example:
/WEB-INF/railo/context/mediaplayer.swf
/WEB-INF/railo/context/mediaplayer.swf.cfm

The mediaplayer.swf.cfm file will serve the mediaplayer.swf file.

Now, my implementation is:
/WEB-INF/railo/context/getfile.cfm
/WEB-INF/railo/context/img/whatever.gif

The getfile.cfm will always be called when requesting static images/js/css, like this: 

That is a request for a cfm file, and works exactly the same as the mediaplayer implementation.
If the mediaplayer works, then the getfile.cfm will work as well, right?

Kind regards,

Paul Klinkenberg

Op 31 mrt 2012, om 22:16 heeft Michael Offner het volgende geschreven:

Yes, a ra file can only contain cfm files. The old version works the same way as your solution but the gifs/pngs are stored in cfm files and can be loaded as include. All this exists and can be enabled by removing a single if statement. But in my tests this was slower and produced a lot of requests. From my point of view we should do a single image that contains all images as collage. Then you define a range in this big image to get a certain image. I have begin with that but run into problems, I have send a mail to the team list and still waiting for feedback ;-)

/Micha

Von meinem iPad gesendet

Am 31.03.2012 um 16:28 schrieb Paul Klinkenbergreply@reply.github.com:

> Hi Micha,
> 
> So, do you mean that a file "something.gif" cannot be inside a .ra file?
> If it can't, then that's the end of the story.
> But otherwise, it will work. I have changed the code so it uses 
> 
> So, all calls for the css/js/img files will still be calls to a .cfm file.
> That will work I suspect?
> 
> Kind regards,
> 
> Paul Klinkenberg
> 
> Op 31 mrt 2012, om 13:10 heeft Michael Offner het volgende geschreven:
> 
> In older versions the images was in external files and still is for older ie (there is a switch), but with extension .cfm. Problem with none cfm files is that they cannot be part of the ra, file and they are not reconized by the server. Because the Railo-context mapping is only know by the cfmlservlet, as long you not install the Railo default servlet. What never is by default.
> In my test the internal way was the best solution. 
> 
> /micha
> 
> Vo mim iPhone gschickt
> 
> Am 31.03.2012 um 11:50 schrieb Paul Klinkenbergreply@reply.github.com:
> 
> > Hi,
> > I was about to start working on the upload functionality for Extensions, when I once again noticed the inline images and inline styles/scripts in the admin, and the .cfm extensions for javascript files.
> > Since the .cfm extensions will always reload on every request, and the inline images make the page unnecessarily heavy, I made some changes.
> > 
> > File getfile.cfm first does a security check for valid file path, and then either returns a 304 Not Modified based on ETag, or the file (img, css, js) contents plus an ETag header.
> > If getfile.cfm is run from /railo-context/getfile.cfm, it will also append 2 caching headers, which should cache the contents for 1 full day. This way, while developing locally, the files will renew, but when released, the files will be renewed only once a day.
> > 
> > Then, I added the following directories, and moved inline styles and javascripts to the respective directories:
> > /railo-admin/img
> > /railo-admin/js
> > /railo-admin/css
> > 
> > Kind regards,
> > 
> > Paul Klinkenberg
> > 
> > You can merge this Pull Request by running:
> > 
> > git pull https://github.com/paulklinkenberg/railo develop-admin-changes
> > 
> > Or you can view, comment on it, or merge it online at:
> > 
> > https://github.com/getrailo/railo/pull/35
> > 
> > -- Commit Summary --
> > - Merge remote branch 'Railo/develop' into develop
> > - Removed unused old css file from the admin
> > - Added getfile.cfm in the admin, which replaces inline images and inline style/javascript.
> > 
> > -- File Changes --
> > 
> > M railo-cfml/railo-admin/admin/admin_layout.cfm (374)
> > D railo-cfml/railo-admin/admin/resources/admin.css.cfm (74)
> > M railo-cfml/railo-admin/admin/web.cfm (140)
> > A railo-cfml/railo-admin/css/style.css (90)
> > A railo-cfml/railo-admin/getfile.cfm (56)
> > A railo-cfml/railo-admin/img/arrow-active.gif (0)
> > A railo-cfml/railo-admin/img/arrow-down.gif (0)
> > A railo-cfml/railo-admin/img/arrow-left.gif (0)
> > A railo-cfml/railo-admin/img/arrow-right.gif (0)
> > A railo-cfml/railo-admin/img/arrow-up.gif (0)
> > A railo-cfml/railo-admin/img/arrow.gif (0)
> > A railo-cfml/railo-admin/img/back-left.png (0)
> > A railo-cfml/railo-admin/img/bg-left.png (0)
> > A railo-cfml/railo-admin/img/box-bg.png (0)
> > A railo-cfml/railo-admin/img/box-del.png (0)
> > A railo-cfml/railo-admin/img/box-left.png (0)
> > A railo-cfml/railo-admin/img/box-right.png (0)
> > A railo-cfml/railo-admin/img/debug_execute.gif (0)
> > A railo-cfml/railo-admin/img/debug_minus.gif (0)
> > A railo-cfml/railo-admin/img/debug_plus.gif (0)
> > A railo-cfml/railo-admin/img/edit.png (0)
> > A railo-cfml/railo-admin/img/input-button.png (0)
> > A railo-cfml/railo-admin/img/input-shadow-error.png (0)
> > A railo-cfml/railo-admin/img/input-shadow.png (0)
> > A railo-cfml/railo-admin/img/left-tab-server.png (0)
> > A railo-cfml/railo-admin/img/left-tab-web.png (0)
> > A railo-cfml/railo-admin/img/left.png (0)
> > A railo-cfml/railo-admin/img/maxl.png (0)
> > A railo-cfml/railo-admin/img/maxr.png (0)
> > A railo-cfml/railo-admin/img/minl.png (0)
> > A railo-cfml/railo-admin/img/minus.png (0)
> > A railo-cfml/railo-admin/img/plus.png (0)
> > A railo-cfml/railo-admin/img/right-tab-server.png (0)
> > A railo-cfml/railo-admin/img/right-tab-web.png (0)
> > A railo-cfml/railo-admin/img/server-back.png (0)
> > A railo-cfml/railo-admin/img/server-bgcolor.gif (0)
> > A railo-cfml/railo-admin/img/server-minus.gif (0)
> > A railo-cfml/railo-admin/img/server-railo-small.png (0)
> > A railo-cfml/railo-admin/img/server-railo.png (0)
> > A railo-cfml/railo-admin/img/shadow-left.gif (0)
> > A railo-cfml/railo-admin/img/shadow-right.gif (0)
> > A railo-cfml/railo-admin/img/tp.gif (0)
> > A railo-cfml/railo-admin/img/web-back.png (0)
> > A railo-cfml/railo-admin/img/web-bgcolor.gif (0)
> > A railo-cfml/railo-admin/img/web-minus.gif (0)
> > A railo-cfml/railo-admin/img/web-railo-small.png (0)
> > A railo-cfml/railo-admin/img/web-railo.png (0)
> > A railo-cfml/railo-admin/js/admin.js (103)
> > A railo-cfml/railo-admin/js/jquery-1.7.2.min.js (4)
> > A railo-cfml/railo-admin/js/jquery.blockUI.js (499)
> > 
> > -- Patch Links --
> > 
> > https://github.com/getrailo/railo/pull/35.patch
> > https://github.com/getrailo/railo/pull/35.diff
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/getrailo/railo/pull/35
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/getrailo/railo/pull/35#issuecomment-4857741
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/getrailo/railo/pull/35#issuecomment-4858762

---

**Comment by micstriit** [comment]
yes mediaplayer is a very very old solution (even older than the
solution from admin we had before the current inline solution),
because the mediaplayer.swf cannot be in the railo-context.ra, i have
to deploy manually in the code.
we have 3 generations of solutions for this problem. the
mediaplayer.swf is generation 1 and the badest solution. then we had
generation 2 that no longer has the content in a none cfm file that
not can be stored in
a ra file. generation 3 is the inline solution and the planned 4th
generation is the image collage solution.

in the admin the call for a pic looks like this
<cfmodule template="img.cfm" src="shadow-left.gif" />
this is the img.cfm
https://github.com/getrailo/railo/blob/3.3/railo-cfml/railo-admin/admin/img.cfm

you can see on line 35 that railo use "oldStyle" for "MSIE" and this
means the pic is embedded as external cfm file, similar to your
getfile.cfm call, but in this case railo has not to load from a
additional external file and the file has not to be deployed manually
because it is in the ra.

i will search for the mail for the 4 generation solution.

/micha

2012/3/31 Paul Klinkenberg
reply@reply.github.com:

**Comment by paulklinkenberg** [comment]
Hi Micha,

Yes, I understand what the problem is: archives can only contain compiled class files. So even if I would rename whatever.gif to whatever.gif.cfm, it would get corrupted when it is compiled.

Just an idea: could we (you) rewrite the archive creation process, to make it possible to also add other content to railo archives? Instead of just looking for cfm/cfc, it would also plain-copy all other found files.
If that can be done, then you not only get rid of a long-time problem with these images, but it would also make Railo archives simpler to use (see for example the long explanation for transforming images here: http://blog.getrailo.com/post.cfm/creating-railo-archives) 
And the internal code and files of the Railo admin would also get much more logical and easy to edit.

The image-crop idea is a very difficult one btw; the designer at my company almost never uses it, because he ran into so much trouble. I actually sent him your email some time ago, and he said it wasn't a good/simple idea :-/

By using the headers Cache-Control, ETag, and Last-Modified, the getfile.cfm would make sure that there is only one call for any image/css/js. That will reduce bandwidth and http calls. I assume that wasn't used when doing "generation 1"?
I think this isn't the baddest solution ;-)

Thanks for the explanation, and I hope we can come up with a good solution!

Kind regards,

Paul Klinkenberg

Op 31 mrt 2012, om 23:06 heeft Michael Offner het volgende geschreven:

yes mediaplayer is a very very old solution (even older than the
solution from admin we had before the current inline solution),
because the mediaplayer.swf cannot be in the railo-context.ra, i have
to deploy manually in the code.
we have 3 generations of solutions for this problem. the
mediaplayer.swf is generation 1 and the badest solution. then we had
generation 2 that no longer has the content in a none cfm file that
not can be stored in
a ra file. generation 3 is the inline solution and the planned 4th
generation is the image collage solution.

in the admin the call for a pic looks like this
<cfmodule template="img.cfm" src="shadow-left.gif" />
this is the img.cfm
https://github.com/getrailo/railo/blob/3.3/railo-cfml/railo-admin/admin/img.cfm

you can see on line 35 that railo use "oldStyle" for "MSIE" and this
means the pic is embedded as external cfm file, similar to your
getfile.cfm call, but in this case railo has not to load from a
additional external file and the file has not to be deployed manually
because it is in the ra.

i will search for the mail for the 4 generation solution.

/micha

2012/3/31 Paul Klinkenberg
reply@reply.github.com:

> Hi Micha,
> 
> I think/hope you understood my solution wrong.
> When I look at the WEB-INF directory of any web context, I see for example:
> /WEB-INF/railo/context/mediaplayer.swf
> /WEB-INF/railo/context/mediaplayer.swf.cfm
> 
> The mediaplayer.swf.cfm file will serve the mediaplayer.swf file.
> 
> Now, my implementation is:
> /WEB-INF/railo/context/getfile.cfm
> /WEB-INF/railo/context/img/whatever.gif
> 
> The getfile.cfm will always be called when requesting static images/js/css, like this: 
> 
> That is a request for a cfm file, and works exactly the same as the mediaplayer implementation.
> If the mediaplayer works, then the getfile.cfm will work as well, right?
> 
> Kind regards,
> 
> Paul Klinkenberg
> 
> Op 31 mrt 2012, om 22:16 heeft Michael Offner het volgende geschreven:
> 
> Yes, a ra file can only contain cfm files. The old version works the same way as your solution but the gifs/pngs are stored in cfm files and can be loaded as include. All this exists and can be enabled by removing a single if statement. But in my tests this was slower and produced a lot of requests. From my point of view we should do a single image that contains all images as collage. Then you define a range in this big image to get a certain image. I have begin with that but run into problems, I have send a mail to the team list and still waiting for feedback ;-)
> 
> /Micha
> 
> Von meinem iPad gesendet
> 
> Am 31.03.2012 um 16:28 schrieb Paul Klinkenbergreply@reply.github.com:
> 
> > Hi Micha,
> > 
> > So, do you mean that a file "something.gif" cannot be inside a .ra file?
> > If it can't, then that's the end of the story.
> > But otherwise, it will work. I have changed the code so it uses 
> > 
> > So, all calls for the css/js/img files will still be calls to a .cfm file.
> > That will work I suspect?
> > 
> > Kind regards,
> > 
> > Paul Klinkenberg
> > 
> > Op 31 mrt 2012, om 13:10 heeft Michael Offner het volgende geschreven:
> > 
> > In older versions the images was in external files and still is for older ie (there is a switch), but with extension .cfm. Problem with none cfm files is that they cannot be part of the ra, file and they are not reconized by the server. Because the Railo-context mapping is only know by the cfmlservlet, as long you not install the Railo default servlet. What never is by default.
> > In my test the internal way was the best solution.
> > 
> > /micha
> > 
> > Vo mim iPhone gschickt
> > 
> > Am 31.03.2012 um 11:50 schrieb Paul Klinkenbergreply@reply.github.com:
> > 
> > > Hi,
> > > I was about to start working on the upload functionality for Extensions, when I once again noticed the inline images and inline styles/scripts in the admin, and the .cfm extensions for javascript files.
> > > Since the .cfm extensions will always reload on every request, and the inline images make the page unnecessarily heavy, I made some changes.
> > > 
> > > File getfile.cfm first does a security check for valid file path, and then either returns a 304 Not Modified based on ETag, or the file (img, css, js) contents plus an ETag header.
> > > If getfile.cfm is run from /railo-context/getfile.cfm, it will also append 2 caching headers, which should cache the contents for 1 full day. This way, while developing locally, the files will renew, but when released, the files will be renewed only once a day.
> > > 
> > > Then, I added the following directories, and moved inline styles and javascripts to the respective directories:
> > > /railo-admin/img
> > > /railo-admin/js
> > > /railo-admin/css
> > > 
> > > Kind regards,
> > > 
> > > Paul Klinkenberg
> > > 
> > > You can merge this Pull Request by running:
> > > 
> > > git pull https://github.com/paulklinkenberg/railo develop-admin-changes
> > > 
> > > Or you can view, comment on it, or merge it online at:
> > > 
> > > https://github.com/getrailo/railo/pull/35
> > > 
> > > -- Commit Summary --
> > > - Merge remote branch 'Railo/develop' into develop
> > > - Removed unused old css file from the admin
> > > - Added getfile.cfm in the admin, which replaces inline images and inline style/javascript.
> > > 
> > > -- File Changes --
> > > 
> > > M railo-cfml/railo-admin/admin/admin_layout.cfm (374)
> > > D railo-cfml/railo-admin/admin/resources/admin.css.cfm (74)
> > > M railo-cfml/railo-admin/admin/web.cfm (140)
> > > A railo-cfml/railo-admin/css/style.css (90)
> > > A railo-cfml/railo-admin/getfile.cfm (56)
> > > A railo-cfml/railo-admin/img/arrow-active.gif (0)
> > > A railo-cfml/railo-admin/img/arrow-down.gif (0)
> > > A railo-cfml/railo-admin/img/arrow-left.gif (0)
> > > A railo-cfml/railo-admin/img/arrow-right.gif (0)
> > > A railo-cfml/railo-admin/img/arrow-up.gif (0)
> > > A railo-cfml/railo-admin/img/arrow.gif (0)
> > > A railo-cfml/railo-admin/img/back-left.png (0)
> > > A railo-cfml/railo-admin/img/bg-left.png (0)
> > > A railo-cfml/railo-admin/img/box-bg.png (0)
> > > A railo-cfml/railo-admin/img/box-del.png (0)
> > > A railo-cfml/railo-admin/img/box-left.png (0)
> > > A railo-cfml/railo-admin/img/box-right.png (0)
> > > A railo-cfml/railo-admin/img/debug_execute.gif (0)
> > > A railo-cfml/railo-admin/img/debug_minus.gif (0)
> > > A railo-cfml/railo-admin/img/debug_plus.gif (0)
> > > A railo-cfml/railo-admin/img/edit.png (0)
> > > A railo-cfml/railo-admin/img/input-button.png (0)
> > > A railo-cfml/railo-admin/img/input-shadow-error.png (0)
> > > A railo-cfml/railo-admin/img/input-shadow.png (0)
> > > A railo-cfml/railo-admin/img/left-tab-server.png (0)
> > > A railo-cfml/railo-admin/img/left-tab-web.png (0)
> > > A railo-cfml/railo-admin/img/left.png (0)
> > > A railo-cfml/railo-admin/img/maxl.png (0)
> > > A railo-cfml/railo-admin/img/maxr.png (0)
> > > A railo-cfml/railo-admin/img/minl.png (0)
> > > A railo-cfml/railo-admin/img/minus.png (0)
> > > A railo-cfml/railo-admin/img/plus.png (0)
> > > A railo-cfml/railo-admin/img/right-tab-server.png (0)
> > > A railo-cfml/railo-admin/img/right-tab-web.png (0)
> > > A railo-cfml/railo-admin/img/server-back.png (0)
> > > A railo-cfml/railo-admin/img/server-bgcolor.gif (0)
> > > A railo-cfml/railo-admin/img/server-minus.gif (0)
> > > A railo-cfml/railo-admin/img/server-railo-small.png (0)
> > > A railo-cfml/railo-admin/img/server-railo.png (0)
> > > A railo-cfml/railo-admin/img/shadow-left.gif (0)
> > > A railo-cfml/railo-admin/img/shadow-right.gif (0)
> > > A railo-cfml/railo-admin/img/tp.gif (0)
> > > A railo-cfml/railo-admin/img/web-back.png (0)
> > > A railo-cfml/railo-admin/img/web-bgcolor.gif (0)
> > > A railo-cfml/railo-admin/img/web-minus.gif (0)
> > > A railo-cfml/railo-admin/img/web-railo-small.png (0)
> > > A railo-cfml/railo-admin/img/web-railo.png (0)
> > > A railo-cfml/railo-admin/js/admin.js (103)
> > > A railo-cfml/railo-admin/js/jquery-1.7.2.min.js (4)
> > > A railo-cfml/railo-admin/js/jquery.blockUI.js (499)
> > > 
> > > -- Patch Links --
> > > 
> > > https://github.com/getrailo/railo/pull/35.patch
> > > https://github.com/getrailo/railo/pull/35.diff
> > > 
> > > ---
> > > 
> > > Reply to this email directly or view it on GitHub:
> > > https://github.com/getrailo/railo/pull/35
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/getrailo/railo/pull/35#issuecomment-4857741
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/getrailo/railo/pull/35#issuecomment-4858762
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/getrailo/railo/pull/35#issuecomment-4861362
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/getrailo/railo/pull/35#issuecomment-4861442

---

**Comment by paulklinkenberg** [comment]
Hi Micha,

I added a new commit, which will work when it is compiled.
Changes:
- added a .cfm extension to js and css files
- removed the plain images I added earlier on
- changed getfile.cfm, so it <cfinclude>s js/css files
- changed getfile.cfm, so it uses the base64 data for the images, which you already created in the admin.

2 reasons to use this solution as the next "generation":
- caching is really strong, so lower bandwidth and less requests
- the code looks far more tidy and is more easy to work with

Hope this helps :-)

Paul

**Comment by micstriit** [comment]
we can check to open ra files for none cfm files, i had this in mind for a
long time, but there was no big pressure, nobody ever asking for this.
But in this case this makes no difference, the only difference between your
getfile.cfm and the img.cfm (old style) is that your solution loads the
image from a file with gif,png... extension and the img.cfm from
png.cfm,gif.cfm ..., we can add your client site caching stuff to the
img.cfm and disable "new style".
i have a deployer app that generates me all this png.cfm,gif.cfm files.
i will build a release that takes your caching for "old style" and disables
"new style", then we can compare this solutions.

i still think the "crop" image is the best solution and the problem i have
can be solved.
even we can solve only for certain browser, we can still run "old style"
and "new style".

/micha

2012/3/31 Paul Klinkenberg <
reply@reply.github.com

**Comment by micstriit** [comment]
sorry i have overlooked this.

i think this is what i mean in my previous mail, it is a mix between
the "old style" solution and your getfile solution.
have you tested the execution time with firebug or something else?
I'm not sure if the client cache really will work, because you call
the getfile.cfm with a "?" and normally this makes the browser
ignoring the response header defintions for caching as far as i know.

perhaps we should call the generated <ext>.cfm files directly and this
files have included the caching definition.
in short term, instead of

we do


and we move the cfheader stuff to the <ext>.cfm files (no includes,
this will make the files more flexible).

the "deployer" that generates this files is here:
https://github.com/getrailo/railo/blob/3.3/railo-cfml/railo-admin/admin/factory/img.cfm

/micha

p.s. i hope you don't get me wrong, your input is very very welcome, i
thank you for the effort you're doing on this.

2012/4/1 Paul Klinkenberg
reply@reply.github.com:

**Comment by paulklinkenberg** [comment]
Hi Micha,

I did some reading and searching on the internet, and the "files with query string won't be cached" idea is meant for old browsers (up untill IE6), with the exception of Squid proxy, which only allows this caching since march 2010 (version 2.7).

Yes, I looked at Firebug in Firefox. It loads the static files once, and then just retrieves them from cache, without new (304) calls to the server. Because the persistent caching headers are set.
While developing, the persistent caching headers are not set btw (just the ETag header), so you will get all 304 calls for images etc.

I don't mind rewriting the code, so it has the caching inside the img.png.cfm file, if you prefer that.
Then, it will look something like this:

<cfsavecontent variable="c">base64-image-data==</cfsavecontent>
<cfif not structKeyExists(application, "browserCachingObj")>
    <cfset application.browserCachingObj = new resources.BrowserCaching() />
</cfif>
<cfset application.browserCachingObj.handleRequest(file=cgi.script_name, returndata=c, convert2Binary=true) />

Just let me know :-)

Kind regards,

Paul Klinkenberg

Op 2 apr 2012, om 08:55 heeft Michael Offner het volgende geschreven:

sorry i have overlooked this.

i think this is what i mean in my previous mail, it is a mix between
the "old style" solution and your getfile solution.
have you tested the execution time with firebug or something else?
I'm not sure if the client cache really will work, because you call
the getfile.cfm with a "?" and normally this makes the browser
ignoring the response header defintions for caching as far as i know.

perhaps we should call the generated <ext>.cfm files directly and this
files have included the caching definition.
in short term, instead of

we do


and we move the cfheader stuff to the <ext>.cfm files (no includes,
this will make the files more flexible).

the "deployer" that generates this files is here:
https://github.com/getrailo/railo/blob/3.3/railo-cfml/railo-admin/admin/factory/img.cfm

/micha

p.s. i hope you don't get me wrong, your input is very very welcome, i
thank you for the effort you're doing on this.

2012/4/1 Paul Klinkenberg
reply@reply.github.com:

> Hi Micha,
> 
> I added a new commit, which will work when it is compiled.
> Changes:
> - added a .cfm extension to js and css files
> - removed the plain images I added earlier on
> - changed getfile.cfm, so it <cfinclude>s js/css files
> - changed getfile.cfm, so it uses the base64 data for the images, which you already created in the admin.
> 
> 2 reasons to use this solution as the next "generation":
> - caching is really strong, so lower bandwidth and less requests
> - the code looks far more tidy and is more easy to work with
> 
> Hope this helps :-)
> 
> Paul
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/getrailo/railo/pull/35#issuecomment-4866602

---

**Comment by paulklinkenberg** [comment]
Hi Micha, could you respond on this one?
I'd like to get it done ;-)

**Comment by micstriit** [comment]
i would prefer to have the caching in the <ext>.cfm files. we just need to
rewrite the deployer a little bit (add the caching stuff) and enable "old
style" in the img.cfm.

sorry for the delay and tnx for your work.

have a nice easter weekend
micha

2012/4/5 Paul Klinkenberg <
reply@reply.github.com

**Comment by paulklinkenberg** [comment]
Hi Micha,

I changed the code again, hopefully to your liking ;-)
To keep the caching code maintainable, I created a cfc for it.
Just pull it in, I'd say ;-)

Paul

**Comment by redler** [comment]
I guess pull requests aren't handled or closed directly in this branch...

**Comment by paulklinkenberg** [comment]
These changes are also included in the new pull request https://github.com/getrailo/railo/pull/59

**Comment by paulklinkenberg** [comment]
This has been pulled meanwhile; closing
